File: java/apts/src/main/java/cn/wjybxx/bigcat/apt/rpc/RpcExporterGenerator.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.bigcat.apt.rpc;

import cn.wjybxx.apt.AbstractGenerator;
import cn.wjybxx.apt.AptUtils;
import cn.wjybxx.apt.BeanUtils;
import com.squareup.javapoet.*;

import javax.lang.model.element.*;
import javax.lang.model.type.TypeKind;
import javax.lang.model.type.TypeMirror;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * @author wjybxx
 * date 2023/4/12
 */
public class RpcExporterGenerator extends AbstractGenerator<RpcServiceProcessor> {

    private static final String varName_registry = "registry";
    private static final String varName_instance = "instance";
    private static final String varName_context = "context";
    private static final String varName_parameter = "parameter";

    private final int serviceId;
    private final List<ExecutableElement> rpcMethods;

    RpcExporterGenerator(RpcServiceProcessor processor, TypeElement typeElement, int serviceId, List<ExecutableElement> rpcMethods) {
        super(processor, typeElement);
        this.serviceId = serviceId;
        this.rpcMethods = rpcMethods;
    }

    @Override
    public void execute() {
        final TypeSpec.Builder typeBuilder = TypeSpec.classBuilder(getServerProxyClassName(typeElement))
                .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
                .addAnnotation(AptUtils.SUPPRESS_UNCHECKED_RAWTYPES)
                .addAnnotation(processorInfoAnnotation)
                .addAnnotation(AptUtils.newSourceFileRefAnnotation(ClassName.get(typeElement)));

        final List<MethodSpec> serverMethodProxyList = new ArrayList<>(rpcMethods.size());
        // 生成代理方法
        for (final ExecutableElement method : rpcMethods) {
            serverMethodProxyList.add(genServerMethodProxy(typeElement, serviceId, method));
        }
        typeBuilder.addMethods(serverMethodProxyList);

        // 生成注册方法
        typeBuilder.addMethod(genRegisterMethod(typeElement, serverMethodProxyList));

        // 写入文件
        AptUtils.writeToFile(typeElement, typeBuilder, elementUtils, messager, filer);
    }

    private String getServerProxyClassName(TypeElement typeElement) {
        return typeElement.getSimpleName().toString() + "Exporter";
    }

    /**
     * 生成注册方法
     * <pre>{@code
     * public static void export(RpcProxyRegistry registry, T instance) {
     *     exportMethod1(registry, instance);
     *     exportMethod2(registry, instance);
     * }
     * </pre>
     *
     * @param serverProxyMethodList 被代理的服务器方法
     */
    private MethodSpec genRegisterMethod(TypeElement typeElement, List<MethodSpec> serverProxyMethodList) {
        MethodSpec.Builder builder = MethodSpec.methodBuilder("export")
                .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
                .returns(TypeName.VOID)
                .addParameter(processor.methodRegistryTypeName, varName_registry)
                .addParameter(TypeName.get(typeElement.asType()), varName_instance);

        // 添加调用
        for (MethodSpec method : serverProxyMethodList) {
            builder.addStatement("$L($L, $L)", method.name, varName_registry, varName_instance);
        }

        return builder.build();
    }

    /**
     * 为某个具体方法生成注册方法，方法分为两类
     * 1. 有返回值的，直接返回方法执行结果; 如果方法签名中包含context，需要传入；
     * <pre>
     * {@code
     * 		private static void exportMethod1(RpcProxyRegistry registry, T instance) {
     * 		    registry.<R>register(10001, (context, parameter) -> {
     * 		        R r = instance.method10001(context, (T) parameter);
     * 		        if (context.isManualReturn()) return;
     * 		        context.sendResult(r);
     *         }
     *     }
     * }
     * {@code
     * 		private static void exportMethod1(RpcProxyRegistry registry, T instance) {
     * 		    registry.<R>register(10001, (context, parameter) -> {
     * 		        IFuture<R> r = instance.method10001(context, (T) parameter);
     * 		        if (context.isManualReturn()) return;
     * 		        context.sendAsyncResult(r);
     *         }
     *     }
     * }
     * </pre>
     * 2. 无返回值的，代理执行完之后直接返回null；如果方法签名中包含context，需要传入
     * <pre>
     * {@code
     *      private static void exportMethod2(RpcProxyRegistry registry, T instance) {
     * 		    registry.<Object>register(10002, (context, parameter) -> {
     * 		        instance.method10001(context, (T) parameter);
     *          }
     *     }
     * }
     * </pre>
     */
    private MethodSpec genServerMethodProxy(TypeElement typeElement, int serviceId, ExecutableElement method) {
        final Map<String, AnnotationValue> annoValueMap = processor.getMethodAnnoValueMap(method);
        final int methodId = processor.getMethodId(method, annoValueMap);
        final MethodSpec.Builder builder = MethodSpec.methodBuilder(getServerProxyMethodName(methodId, method))
                .addModifiers(Modifier.PRIVATE, Modifier.STATIC)
                .returns(TypeName.VOID)
                .addParameter(processor.methodRegistryTypeName, varName_registry)
                .addParameter(TypeName.get(typeElement.asType()), varName_instance);
        // 拷贝泛型参数
        AptUtils.copyTypeVariables(builder, method);
        // 注册方法代理
        builder.addCode(genMethodProxy(serviceId, method, methodId, annoValueMap).build());
        // 注册切面数据
        String customData = processor.getCustomData(method, annoValueMap);
        if (customData != null) {
            builder.addStatement("$L.setProxyData($L, $L, $S)", varName_registry, serviceId, methodId, customData);
        }
        return builder.build();
    }

    /** 生成方法代理 */
    private CodeBlock.Builder genMethodProxy(int serviceId, ExecutableElement method, int methodId,
                                             Map<String, AnnotationValue> annoValueMap) {
        CodeBlock.Builder codeBuilder = CodeBlock.builder();
        // registry -- 传入泛型参数，可以避免不必要的类型转换
        TypeMirror rpcReturnType = processor.rpcReturnType(method, true);
        codeBuilder.beginControlFlow("$L.<$T>register($L, $L, (context, $L) ->",
                varName_registry, rpcReturnType, serviceId, methodId, varName_parameter);
        // 可变性设置
        if (processor.isResultSharable(method, annoValueMap)) {
            codeBuilder.addStatement("context.setSharable(true)");
        }
        if (processor.isManualReturn(method, annoValueMap)) {
            codeBuilder.addStatement("context.setManualReturn(true)");
        }
        // 执行方法调用 -- 这里测试方法的直接返回值
        if (method.getReturnType().getKind() == TypeKind.VOID) {
            StringBuilder format = new StringBuilder(32);
            List<Object> params = new ArrayList<>(4);
            genInvokeStatement(method, format, params);
            codeBuilder.addStatement(format.toString(), params.toArray()); // 需要ToArray
        } else {
            StringBuilder format = new StringBuilder(32);
            List<Object> params = new ArrayList<>(4);
            {
                format.append("$T tempR = ");
                params.add(TypeName.get(method.getReturnType()));
            }
            genInvokeStatement(method, format, params);
            codeBuilder.addStatement(format.toString(), params.toArray()); // 需要ToArray

            codeBuilder.addStatement("if (context.isManualReturn()) return");
            if (processor.isFuture(method.getReturnType())) {
                codeBuilder.addStatement("context.sendAsyncResult(tempR)");
            } else {
                codeBuilder.addStatement("context.sendResult(tempR)");
            }
        }
        codeBuilder.unindent(); // endControlFlow会拼入空格...
        codeBuilder.addStatement("})");
        return codeBuilder;
    }

    /**
     * 获取代理方法的名字
     */
    private static String getServerProxyMethodName(int methodId, ExecutableElement method) {
        // 加上methodId防止重复
        return "_export" + BeanUtils.firstCharToUpperCase(method.getSimpleName().toString()) + "_" + methodId;
    }

    /**
     * 生成方法调用代码，没有分号和换行符。
     * {@code instance.rpcMethod(a, b, c)}
     */
    private void genInvokeStatement(ExecutableElement method, StringBuilder format, List<Object> params) {
        // 调用方法
        format.append("$L.$L(");
        params.add(varName_instance);
        params.add(method.getSimpleName().toString());

        // 去除context -- Context的类型转换已在上面统一处理
        List<? extends VariableElement> parameters = method.getParameters();
        if (parameters.size() > 0 && processor.isContext(parameters.get(0).asType())) {
            format.append("context");
            if (parameters.size() > 1) {
                format.append(", ");
            }
            parameters = method.getParameters().subList(1, parameters.size());
        }
        // 方法参数已限定为最多1个，Object向下转换
        if (parameters.size() > 0) {
            VariableElement variableElement = parameters.get(0);
            final TypeName parameterTypeName = TypeName.get(variableElement.asType());

            format.append("($T) $L");
            params.add(parameterTypeName);
            params.add(varName_parameter);
        }
        format.append(")");
    }
}

File: csharp/Wjybxx.BigCat.Apt/src/RpcExporterGenerator.cs
#region LICENSE

// Copyright 2025 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System.Collections.Generic;
using System.Collections.Immutable;
using System.Text;
using Microsoft.CodeAnalysis;
using Wjybxx.Commons.Apt;
using Wjybxx.Commons.Poet;

namespace Wjybxx.BigCat.Apt
{
/// <summary>
/// 生成服务器的方法导出代码(lambda)
/// </summary>
public class RpcExporterGenerator
{
    private const string varName_registry = "registry";
    private const string varName_instance = "instance";
    private const string varName_context = "context";
    private const string varName_parameter = "parameter";

    private readonly RpcServiceProcessor processor;
    private readonly INamedTypeSymbol typeSymbol;

    private readonly int serviceId;
    private readonly List<IMethodSymbol> rpcMethods;
    private readonly ClassName serviceTypeName;

    public RpcExporterGenerator(RpcServiceProcessor processor, INamedTypeSymbol typeSymbol,
                                int serviceId, List<IMethodSymbol> rpcMethods) {
        this.processor = processor;
        this.typeSymbol = typeSymbol;
        this.serviceId = serviceId;
        this.rpcMethods = rpcMethods;
        this.serviceTypeName = (ClassName)AptUtils.ParseType(typeSymbol);
    }

    public TypeSpec Execute() {
        TypeSpec.Builder typeBuilder = TypeSpec.NewClassBuilder(GetServerProxyClassName(typeSymbol))
            .AddModifiers(Modifiers.Public | Modifiers.Sealed)
            .AddAttribute(processor.processorInfoAnnotation)
            .AddAttribute(AptUtils.NewSourceFileRefAnnotation(AptUtils.ParseType(typeSymbol)));

        List<MethodSpec> serverMethodProxyList = new(rpcMethods.Count);
        // 生成代理方法
        foreach (IMethodSymbol method in rpcMethods) {
            serverMethodProxyList.Add(GenServerMethodProxy(method));
        }
        typeBuilder.AddMethods(serverMethodProxyList);

        // 生成注册方法
        typeBuilder.AddMethod(GenRegisterMethod(serverMethodProxyList));
        return typeBuilder.Build();
    }

    private static string GetServerProxyClassName(INamedTypeSymbol typeSymbol) {
        return typeSymbol.Name + "Exporter";
    }

    /// <summary>
    /// 生成注册方法
    /// 
    /// <![CDATA[
    /// public static void export(RpcProxyRegistry registry, T instance) {
    ///     exportMethod1(registry, instance);
    ///     exportMethod2(registry, instance);
    /// }
    /// ]]>
    /// </summary>
    /// <param name="serverProxyMethodList"></param>
    /// <returns></returns>
    private MethodSpec GenRegisterMethod(List<MethodSpec> serverProxyMethodList) {
        MethodSpec.Builder builder = MethodSpec.NewMethodBuilder("Export")
            .AddModifiers(Modifiers.Public | Modifiers.Static)
            .Returns(TypeName.VOID)
            .AddParameter(processor.typeName_MethodRegistry, varName_registry)
            .AddParameter(serviceTypeName, varName_instance);
        // 添加调用
        foreach (MethodSpec method in serverProxyMethodList) {
            builder.codeBuilder.AddStatement("$L($L, $L)", method.name, varName_registry, varName_instance);
        }
        return builder.Build();
    }

    /// <summary>
    /// 
    /// </summary>
    /// <param name="method"></param>
    /// <returns></returns>
    private MethodSpec GenServerMethodProxy(IMethodSymbol method) {
        AttributeData annoValueMap = processor.GetMethodAnnoValueMap(method)!;
        int methodId = processor.GetMethodId(method, annoValueMap);
        MethodSpec.Builder builder = MethodSpec.NewMethodBuilder(GetServerProxyMethodName(methodId, method))
            .AddModifiers(Modifiers.Private | Modifiers.Static)
            .Returns(TypeName.VOID)
            .AddParameter(processor.typeName_MethodRegistry, varName_registry)
            .AddParameter(serviceTypeName, varName_instance);
        // 拷贝泛型参数
        AptUtils.CopyTypeParameters(builder, method);
        // 注册方法代理
        builder.codeBuilder.Add(GenMethodProxy(method, methodId, annoValueMap).Build());
        // 注册切面数据
        string customData = processor.GetCustomData(method, annoValueMap);
        if (customData != null) {
            builder.codeBuilder.AddStatement("$L.setProxyData($L, $L, $S)", varName_registry, serviceId, methodId, customData);
        }
        return builder.Build();
    }

    /** 生成方法代理 */
    private CodeBlock.Builder GenMethodProxy(IMethodSymbol method, int methodId,
                                             AttributeData annoValueMap) {
        CodeBlock.Builder codeBuilder = CodeBlock.NewBuilder();
        // registry -- 传入泛型参数，可以避免不必要的类型转换；要使用ref必须显式声明context的类型
        TypeName rpcReturnTypeName = AptUtils.ParseType(processor.RpcReturnType(method));
        TypeName rpcContextTypeName = processor.typeName_Context_T.WithTypeArguments(rpcReturnTypeName);
        codeBuilder.BeginControlFlow("$L.Register<$T>($L, $L, (ref $T context, object $L) =>",
            varName_registry, rpcReturnTypeName, serviceId, methodId,
            rpcContextTypeName, varName_parameter);
        // 可变性设置
        if (processor.IsResultSharable(method, annoValueMap)) {
            codeBuilder.AddStatement("context.IsSharable = true");
        }
        if (processor.IsManualReturn(method, annoValueMap)) {
            codeBuilder.AddStatement("context.IsManualReturn = true");
        }
        // 执行方法调用 -- 这里测试方法的直接返回值
        if (method.ReturnsVoid) {
            StringBuilder format = new StringBuilder(32);
            List<object> arguments = new(4);
            genInvokeStatement(method, format, arguments);
            codeBuilder.AddStatement(format.ToString(), arguments.ToArray()); // 需要ToArray
        } else {
            StringBuilder format = new StringBuilder(32);
            List<object> arguments = new(4);
            {
                format.Append("$T tempR = ");
                arguments.Add(AptUtils.ParseType(method.ReturnType));
            }
            genInvokeStatement(method, format, arguments);
            codeBuilder.AddStatement(format.ToString(), arguments.ToArray()); // 需要ToArray

            codeBuilder.AddStatement("if (context.IsManualReturn) return");
            if (processor.IsFuture(method.ReturnType)) {
                codeBuilder.AddStatement("context.SendAsyncResult(tempR)");
            } else {
                codeBuilder.AddStatement("context.SendResult(tempR)");
            }
        }
        codeBuilder.Unindent(); // endControlFlow会拼入空格...
        codeBuilder.AddStatement("})");
        return codeBuilder;
    }

    /**
     * 获取代理方法的名字
     */
    private static string GetServerProxyMethodName(int methodId, IMethodSymbol method) {
        // 加上methodId防止重复
        return "_Export" + Util.FirstCharToUpperCase(method.Name) + "_" + methodId;
    }

    /**
     * 生成方法调用代码，没有分号和换行符。
     * {@code instance.rpcMethod(a, b, c)}
     */
    private void genInvokeStatement(IMethodSymbol method, StringBuilder format, List<object> arguments) {
        // 调用方法
        format.Append("$L.$L(");
        arguments.Add(varName_instance);
        arguments.Add(method.Name);

        // 去除context -- Context的类型转换已在上面统一处理
        ImmutableArray<IParameterSymbol> parameters = method.Parameters;
        if (parameters.Length > 0 && processor.IsContext(parameters[0].Type)) {
            // context需要使用ref/in修饰
            RefKind refKind = parameters[0].RefKind;
            if (refKind == RefKind.Ref) {
                format.Append("ref ");
            } else if (refKind == RefKind.In) {
                format.Append("in ");
            }
            format.Append("context");
            if (parameters.Length > 1) {
                format.Append(", ");
            }
            parameters = parameters.RemoveAt(0);
        }
        // 方法参数已限定为最多1个，Object向下转换
        if (parameters.Length > 0) {
            IParameterSymbol variableElement = parameters[0];
            TypeName parameterTypeName = AptUtils.ParseType(variableElement.Type);

            format.Append("($T) $L");
            arguments.Add(parameterTypeName);
            arguments.Add(varName_parameter);
        }
        format.Append(")");
    }
}
}